### PR TITLE
#4079 support timeframe property in lighthouse service

### LINKED ISSUE #4079: lighthouse: Allow timeframe to be passed via cloudevent
Follow up of https://github.com/keptn/keptn/issues/4075

As a user, I want to be able to specify a timeframe for an evaluation, e.g.:
```
            - name: "evaluation"
              triggeredAfter: "10m"
              properties:
                timeframe: "5m"
```

This needs to be implemented in lighthouse-service, e.g., here:
https://github.com/keptn/keptn/blob/94dfd0a13961ae89d276a892612a2647097363bb/lighthouse-service/event_handler/start_evaluation_handler.go#L145-L152

Note: `start` should be inferred from the current cloudevent evaluation.triggered timestamp. `end` should be calculated as `start - timeframe`.
This is currently handled in API and should be refactored accordingly:
https://github.com/keptn/keptn/blob/069dd0f5c7b6f37a3737f4c0c9c7cf07a801b039/api/handlers/evaluation.go#L84-L131

# Definition of Done

- [ ] lighthouse-service accepts `evaluation.timeframe` in addition to `test.start`, `test.end`, `evaluation.start` and `evaluation.end`
- [ ] keptn/spec and keptn/go-utils `evaluation.triggered` cloudevent needs to be adapted to also include `timeframe`
- [ ] api-service refactored/changed to avoid duplicated code between lighthouse and api-service.

### LINKED ISSUE #4091: Shipyard controller: triggeredID of <stage>.<sequence>.finished events not set properly
When executing multiple iterations of a sequence, the `<stage>.<sequence>.finished` of all iterations reference the id of the first `<stage>.<sequence>.triggered` event. This prevents the bridge from correlating the .finished events to their correct .triggered events which leaves the sequence execution in a pending state, as shown in the screenshot below:

**Current Behavior:**

Sequence execution is displayed as "running/waiting", even though it is already finished

Screenshot 2021-05-18 at 08 48 12

**Involved CloudEvents:**

First remediation.triggered event:

```
{
  "data": {
    "project": "sockshop",
    "service": "carts",
    "stage": "production"
  },
  "id": "2c563526-f6b1-445b-a044-55a33b903d6c",   // <<< this should be referenced by the first remediation.triggered event
  "source": "dynatrace-service",
  "specversion": "1.0",
  "time": "2021-05-18T06:10:25.958Z",
  "type": "sh.keptn.event.production.remediation.triggered",
  "shkeptncontext": "7058be20-179d-49e5-aa54-8c8ddfb90af2"
}
```

**First remediation.finished event:**

```
{
  "data": {
    "project": "sockshop",
    "result": "fail",
    "service": "carts",
    "stage": "production",
    "status": "succeeded"
  },
  "id": "4f62ccf7-6608-484c-b076-8112e106500b",
  "source": "shipyard-controller",
  "specversion": "1.0",
  "time": "2021-05-18T06:20:29.072Z",
  "type": "sh.keptn.event.production.remediation.finished",
  "shkeptncontext": "7058be20-179d-49e5-aa54-8c8ddfb90af2",
  "shkeptnspecversion": "0.2.0",
  "triggeredid": "2c563526-f6b1-445b-a044-55a33b903d6c"
}
```

**Second remediation.triggered event:**

```
{
  "data": {
    "project": "sockshop",
    "service": "carts",
    "stage": "production"
  },
  "id": "71f30038-0fc6-4d6c-8f97-2927734bc775", // <<< this should be referenced by the second remediation.finished event
  "source": "shipyard-controller",
  "specversion": "1.0",
  "time": "2021-05-18T06:20:29.076Z",
  "type": "sh.keptn.event.production.remediation.triggered",
  "shkeptncontext": "7058be20-179d-49e5-aa54-8c8ddfb90af2",
  "shkeptnspecversion": "0.2.0"
}
```

**Second remediation.finished event:**

```
{
  "data": {
    "message": "unable to get next action from remediation.yaml",
    "project": "sockshop",
    "result": "fail",
    "service": "carts",
    "stage": "production",
    "status": "succeeded"
  },
  "id": "31d1f74f-aa0e-4e5c-9898-7de6422cad87",
  "source": "shipyard-controller",
  "specversion": "1.0",
  "time": "2021-05-18T06:20:29.213Z",
  "type": "sh.keptn.event.production.remediation.finished",
  "shkeptncontext": "7058be20-179d-49e5-aa54-8c8ddfb90af2",
  "shkeptnspecversion": "0.2.0",
  "triggeredid": "2c563526-f6b1-445b-a044-55a33b903d6c" // <<< this is the id of the first remediation.triggered event. However, this should be the id of the second remediation.triggered event instead
}
```

**Desired Behavior:**

Sequence execution is displayed as "finished" when there are no open `<stage>.<sequence>.triggered` events anymore:

Screenshot 2021-05-18 at 08 38 59

### PR DESCRIPTION
**Pull request by bacherfl**
Closes #4079
Closes #4091

This PR also fixes a problem that caused the `triggeredid` of a `<stage>.<sequence>.finished`event to not be set properly when there are multiple iterations of the same sequence.

Signed-off-by: Florian Bacher <florian.bacher@dynatrace.com>

Remediation with one iteration:

Screenshot 2021-05-18 at 08 20 21

Note: The `warning` result is caused by a modification I did to the SLO.yaml in order to provoke another iteration

Remediation with two iterations:

Screenshot 2021-05-18 at 08 38 59

Integration test run: https://github.com/keptn/keptn/actions/runs/852286068

### CODE REVIEW

File: lighthouse-service/event_handler/start_evaluation_handler.go [hunk 1]
```diff
@@ -143,8 +144,17 @@ func (eh *StartEvaluationHandler) sendEvaluationFinishedWithErrorEvent(start, en
...
+	if (e.Evaluation.Start != "" && e.Evaluation.End != "") || (e.Evaluation.Timeframe != "") {
+		params := timeutils.GetStartEndTimeParams{
+			StartDate: e.Evaluation.Start,
+			EndDate:   e.Evaluation.End,
+			Timeframe: e.Evaluation.Timeframe,
+		}
+		start, end, err := timeutils.GetStartEndTime(params)
+		if err != nil {
+			return "", "", err
+		}
```
Comment (christian-kreuzberger-dtx) [inline lighthouse-service/event_handler/start_evaluation_handler.go:156]: What happens when timeframe is set, but start and end are not set?

I think in issue #4079 I got it wrong, what I meant was that
`end = evaluation.triggered.time`, and `start = end - timeframe`

Comment (bacherfl) [reply lighthouse-service/event_handler/start_evaluation_handler.go:156]: when only the time frame is set, we adapt the same logic we did previously in the evaluation endpoint: `end = time.Now()`, `start = time.Now() - timeframe`

Comment (christian-kreuzberger-dtx) [reply lighthouse-service/event_handler/start_evaluation_handler.go:156]: The big difference here is that the evaluation endpoint is called in a specific moment.

The cloudevent is sent at a specific moment, but the time it is retrieved by lighthouse could be a different time.
Hence I would propose to not use `time.Now()`, but the actual cloudevents triggered time (when was the evaluation triggered). 

I know it doesn't make much difference in our current setup (maybe a couple of milliseconds), but using the cloudevents time would make the behaviour reproducible and deterministic.
Think about re-triggering an evaluation.

**Review by christian-kreuzberger-dtx — APPROVED** [review]
LGTM

**Pull request merged**